### PR TITLE
refactor(misconf): Merge `trivy-iac` into Trivy

### PR DESCRIPTION
**Pull request by simar7**
## Description

Merges `trivy-iac` into Trivy.

## Related issues
- Close  https://github.com/aquasecurity/trivy/issues/5626

## Related PRs
- [x] https://github.com/aquasecurity/trivy-policies/pull/39
- [x] https://github.com/aquasecurity/trivy-aws/pull/55
- [x] https://github.com/aquasecurity/defsec/pull/1499

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [ ] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -416,3 +418,9 @@ replace oras.land/oras-go => oras.land/oras-go v1.2.4-0.20230801060855-932dd06d3
 // testcontainers-go has a bug with versions v0.25.0 and v0.26.0
 // ref: https://github.com/testcontainers/testcontainers-go/issues/1782
 replace github.com/testcontainers/testcontainers-go => github.com/testcontainers/testcontainers-go v0.23.0
+
+replace github.com/aquasecurity/trivy-policies => /Users/simarpreetsingh/repos/trivy-policies
+
+replace github.com/aquasecurity/trivy-aws => /Users/simarpreetsingh/repos/trivy-aws
+
+replace github.com/aquasecurity/defsec => /Users/simarpreetsingh/repos/defsec
```
Comment (simar7) [inline go.mod:426]: Update when https://github.com/aquasecurity/trivy-policies/pull/39 and https://github.com/aquasecurity/trivy-aws/pull/55 have merged.

File: cmd/schema/main.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
+package main
```
Comment (nikpivkin) [inline cmd/schema/main.go:1]: JSON schema is generated in defsec

File: pkg/rego/build.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
+package rego
```
Comment (nikpivkin) [inline pkg/rego/build.go:1]: The `rego` package has been moved to defsec

File: pkg/rules/register.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package rules
```
Comment (nikpivkin) [inline pkg/rules/register.go:1]: The `rules` package has been moved to defsec

File: internal/rules/register.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
+package rules
```
Comment (nikpivkin) [inline internal/rules/register.go:1]: The `rules` package has been moved to defsec

File: magefiles/magefile.go [hunk 6]
```diff
@@ -425,3 +425,8 @@ func installed(cmd string) bool {
 	_, err := exec.LookPath(cmd)
 	return err == nil
 }
+
+// Schema generates the schema for cloud scanning
+func Schema() error {
```
Comment (nikpivkin) [inline magefiles/magefile.go:430]: same as above

**Comment by simar7** [comment]
@nikpivkin as you mentioned, I removed the packages that exist in defsec today so we can properly make another PR to bring them into Trivy later.

**Comment by simar7** [comment]
@DmitriyLewen the errors in the CI seem unrelated to this PR. Is it because we have to update `tinygo` version after the bump to Go 1.21?
image
image

**Comment by DmitriyLewen** [comment]
Hello @simar7 
We have not received this error before.

I tried to reproduce your error locally, but tests for your PR work correctly:
```bash
➜  trivy git:(merge-trivy-iac-2) ✗ tinygo version
tinygo version 0.29.0 darwin/amd64 (using go version go1.21.4 and LLVM version 15.0.0)

➜  trivy git:(merge-trivy-iac-2) ✗ mage test:module   
...
=== RUN   TestModule
=== RUN   TestModule/spring4shell_jre_8,_severity_update
=== RUN   TestModule/spring4shell_jre_11,_no_severity_update
--- PASS: TestModule (6.04s)
    --- PASS: TestModule/spring4shell_jre_8,_severity_update (2.68s)
    --- PASS: TestModule/spring4shell_jre_11,_no_severity_update (3.20s)
PASS
ok      github.com/aquasecurity/trivy/integration       7.578s

```

I also pulled changes (bump to 1.21) to my PR - #5630. 
Tests are successful.

I see you update/add some dependencies.
Perhaps there is a conflict?

**Comment by simar7** [comment]
> Hello @simar7 We have not received this error before.
> 
> I tried to reproduce your error locally, but tests for your PR work correctly:
> 
> ```shell
> ➜  trivy git:(merge-trivy-iac-2) ✗ tinygo version
> tinygo version 0.29.0 darwin/amd64 (using go version go1.21.4 and LLVM version 15.0.0)
> 
> ➜  trivy git:(merge-trivy-iac-2) ✗ mage test:module   
> ...
> === RUN   TestModule
> === RUN   TestModule/spring4shell_jre_8,_severity_update
> === RUN   TestModule/spring4shell_jre_11,_no_severity_update
> --- PASS: TestModule (6.04s)
>     --- PASS: TestModule/spring4shell_jre_8,_severity_update (2.68s)
>     --- PASS: TestModule/spring4shell_jre_11,_no_severity_update (3.20s)
> PASS
> ok      github.com/aquasecurity/trivy/integration       7.578s
> ```
> 
> I also pulled changes (bump to 1.21) to my PR - #5630. Tests are successful.
> 
> I see you update/add some dependencies. Perhaps there is a conflict?

I see - it's strange indeed. I will try to rebase and see what happens. Thanks for looking into it.

**Comment by nikpivkin** [comment]
I found the following in the tinygo [documentation](https://tinygo.org/docs/reference/lang-support/stdlib/#osuser) on the supported packages page: 
> os/user
> The compiler gave the following error when this package was imported:
>\# os/user
>../../../../../../usr/local/go1.20/src/os/user/cgo_lookup_cgo.go:14:6: not implemented: build constraints in #cgo line

**Comment by simar7** [comment]
> I found the following in the tinygo [documentation](https://tinygo.org/docs/reference/lang-support/stdlib/#osuser) on the supported packages page:
> 
> > os/user
> > The compiler gave the following error when this package was imported:
> > # os/user
> > ../../../../../../usr/local/go1.20/src/os/user/cgo_lookup_cgo.go:14:6: not implemented: build constraints in #cgo line
> 

Interesting find, but we don't use cgo in Trivy or is it a dependency of Trivy that does so (maybe tinygo?) we might have to look through the diff of the changed dependencies [here](https://github.com/aquasecurity/trivy/pull/5636/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6)
What do the entries on that page imply? Are they just outputs of tinygo running against certain code?

> But I used `go mod graph` and couldn't find a package that used that.
> 
> Also I get the following error if I disable CGO (which does not happen in main):
> 
> ```shell
> CGO_ENABLED=0 go generate pkg/module/testdata/analyzer/analyzer.go
> # golang.org/x/sys/unix
> ../../../../../../go/pkg/mod/golang.org/x/sys@v0.14.0/unix/syscall_linux.go:1893:64: undefined: syscall.Rlimit
> ../../../../../../go/pkg/mod/golang.org/x/sys@v0.14.0/unix/syscall_linux.go:1898:50: undefined: syscall.Rlimit
> ../../../../../../go/pkg/mod/golang.org/x/sys@v0.14.0/unix/syscall_linux.go:1898:79: undefined: syscall.Rlimit
> ../../../../../../go/pkg/mod/golang.org/x/sys@v0.14.0/unix/syscall_unix.go:605:17: undefined: syscall.Setrlimit
> ../../../../../../go/pkg/mod/golang.org/x/sys@v0.14.0/unix/syscall_unix.go:605:47: undefined: syscall.Rlimit
> pkg/module/testdata/analyzer/analyzer.go:1: running "tinygo": exit status 1
> ```

Hmm.. I'm also not sure. So far I've tried the following with no success:

1. Downgrade Trivy version to 1.20
2. Upgrade tinygo to v0.30.0
3. Disabling CGO intentionally by passing in `CGO_ENABLED=0` as an argument to the mage module test. Using `sh.RunWith()` [here](https://github.com/aquasecurity/trivy/blob/b970f8c1280d2632ff086204ac803bca68a4fed4/magefiles/magefile.go#L226) I got this idea from [here](https://github.com/tinygo-org/tinygo/issues/3044)

I'm not able to reproduce this issue locally either. I am clearing go test cache between tests via `go clean -testcache`.

**Comment by nikpivkin** [comment]
@simar7 I tried [re-migrating](https://github.com/aquasecurity/trivy/pull/5692) trivy-iac and locally all my tests passed. CI failed due to linter issues.

**Comment by simar7** [comment]
Closing in favor of https://github.com/aquasecurity/trivy/pull/5692 as this one has something up with the CI actions.

**Pull request closed without merging**